lib/loglimit/loglimit.go

```diff
@@ -0,0 +1,198 @@
+You may obtain a copy of the License at
+
+	http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+package loglimit
+
+import (
+	"context"
+	"fmt"
+	"strings"
+	"time"
+
+	"github.com/gravitational/trace"
+	"github.com/jonboulle/clockwork"
+	log "github.com/sirupsen/logrus"
+)
+
+const (
+	// timeWindow is the time window over which logs are deduplicated.
+	timeWindow = time.Minute
+	// timeWindowCleanupInterval is the interval between cleanups of time
+	// windows that have already ended.
+	// Since the time window is set to 1 minute and the cleanup interval
+	// is set to 10 seconds, time windows can be in fact slightly larger
+	// than 1 minute (i.e., 1 minute and 10 seconds, in the worst case).
+	timeWindowCleanupInterval = 10 * time.Second
+)
+
+// Config contains the log limiter config.
+type Config struct {
+	// LogSubstrings contains a list of substrings belonging to the
+	// logs that should be deduplicated.
+	LogSubstrings []string
+	// ChannelSize is the size of the channel used to send messages
+	// to the log limiter.
+	ChannelSize int
+	// Clock is a clock to override in tests, set to real time clock
+	// by default.
+	Clock clockwork.Clock
+}
+
+// checkAndSetDefaults verifies configuration and sets defaults
+func (c *Config) checkAndSetDefaults() error {
+	if c.LogSubstrings == nil {
+		return trace.BadParameter("missing parameter LogSubstrings")
+	}
+	if c.ChannelSize < 0 {
+		return trace.BadParameter("ChannelSize must be at least 0")
+	}
+	if c.Clock == nil {
+		c.Clock = clockwork.NewRealClock()
+	}
+	return nil
+}
+
+// LogLimiter deduplicates logs over a certain time window.
+type LogLimiter struct {
+	// config is the log limiter config.
+	config Config
+	// entryCh is used to send log entries to the log limiter.
+	entryCh chan *entryInfo
+	// windows is a mapping from log substring to an active
+	// time window.
+	windows map[string]*entryInfo
+	// cleanupCh is used in tests to trigger `cleanup`.
+	cleanupCh chan chan struct{}
+}
+
+// entryInfo contains information about a certain log entry.
+// This is information is used to deduplicate a log entry
+// reported within a certain time window.
+type entryInfo struct {
+	// entry is the logger entry.
+	entry *log.Entry
+	// level is the log level at which this log entry should be logged.
+	level log.Level
+	// message is the message message.
+	message string
+	// time is the at which the log reported.
+	time time.Time
+	// occurrences are the occurrences of logs entries (that share
+	// the same log substring) within a time window.
+	occurrences int
+}
+
+// New creates a new log limiter.
+func New(config Config) (*LogLimiter, error) {
```

Comment:
Suggestion: how about moving the `logLimiter.Run()` call inside New, and making `Run` private, so there's less chance of API misuse?

```suggestion
// New creates a new log limiter.
// The ctx passed to New is used to stop the [LogLimiter]. Any calls to
// [LogLimiter.Log] after `ctx` is done will either silently fail or hang the
// program.
func New(ctx context.Context, config Config) (*LogLimiter, error) {
```

Reply:
I'm not sure about this. I can see this pattern (`New` followed by `go Run`/`Start`/`Serve`/...) being used in [heartbeats](https://github.com/gravitational/teleport/blob/e861946ea1b1a6aec2092329ebceb9d68c2cd730/lib/srv/db/server.go#L576), [usage reporting](https://github.com/gravitational/teleport/blob/e861946ea1b1a6aec2092329ebceb9d68c2cd730/lib/teleterm/daemon/daemon.go#L56), [multiplexing](https://github.com/gravitational/teleport/blob/e861946ea1b1a6aec2092329ebceb9d68c2cd730/lib/service/service.go#L1677-L1688), [TLS](https://github.com/gravitational/teleport/blob/e861946ea1b1a6aec2092329ebceb9d68c2cd730/lib/service/service.go#L1711), [tunnel manager](https://github.com/gravitational/teleport/blob/e861946ea1b1a6aec2092329ebceb9d68c2cd730/lib/service/service.go#L3814), [cert reloading](https://github.com/gravitational/teleport/blob/e861946ea1b1a6aec2092329ebceb9d68c2cd730/lib/service/service.go#L4269), [agent pool](https://github.com/gravitational/teleport/blob/e861946ea1b1a6aec2092329ebceb9d68c2cd730/lib/service/service.go#L2394), [reverse tunnels](https://github.com/gravitational/teleport/blob/e861946ea1b1a6aec2092329ebceb9d68c2cd730/lib/service/service.go#L3492), and others.

Now that I'm noticing this, it may have been a mistake to move the start of the log limiter goroutine to `Mux.New` (in 3554daa6db947bd222191b45f922fa428c044673) as the expectation may be that only `Run`/`Start`/`Serve` do start the work.

WDYT?

Reply:
Yep, it comes across often, probably copied from one place to another. Feel free to skip the suggestion if you think it's best.

Reply:
d8f947a42cc03ec2a3196a5883afae1f5582ef67 moves the goroutine spawning to `loglimit.New`.